File: leetcode/894.py
"""
Runtime: 240 ms, faster than 34.10% of Python3 online submissions for All Possible Full Binary Trees.
Memory Usage: 23.1 MB, less than 27.46% of Python3 online submissions for All Possible Full Binary Trees.
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def allPossibleFBT(self, n: int) -> List[TreeNode]:
        if not n % 2:
            return []
        if n == 1:
            return [TreeNode()]

        m, n_orig = 0, n  # m + n = n_orig
        res = []
        while n >= 2:
            n -= 2
            m = n_orig - n - 1
            dp_m = self.allPossibleFBT(m)
            dp_n = self.allPossibleFBT(n)
            res += [
                TreeNode(left=t_m, right=t_n) for t_m in dp_m for t_n in dp_n
            ]
            # for t_m in dp_m:
            #    for t_n in dp_n:
            #        r = TreeNode()
            #        r.left, r.right = t_m, t_n
            #        res.append(r)

        return res

File: leetcode/894.cpp
// Runtime: 109 ms, faster than 73.84% of C++ online submissions for All Possible Full Binary Trees.
// Memory Usage: 29.2 MB, less than 59.02% of C++ online submissions for All Possible Full Binary Trees.
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
 private:
  std::map<int, vector<TreeNode *>> n_to_treenode_;

 public:
  vector<TreeNode *> allPossibleFBT(int n) {
    if (n_to_treenode_.find(n) != n_to_treenode_.end()) {
      return n_to_treenode_[n];
    }
    // we can build the FBT recursively
    // for each FBT of size n-1, we build all size n FBT
    // by making the n-1 a left or right subtree
    if (n < 1) {
      return vector<TreeNode *>({});
    }
    if (n == 1) {
      return vector<TreeNode *>({new TreeNode});
    }

    vector<TreeNode *> res = {};

    for (int m = 1; m < n; m += 2) {
      vector<TreeNode *> subtrees_l_ptr = allPossibleFBT(m);
      vector<TreeNode *> subtrees_r_ptr = allPossibleFBT(n - m - 1);
      for (auto &subtree_l_ptr : subtrees_l_ptr) {
        for (auto &subtree_r_ptr : subtrees_r_ptr) {
          TreeNode *t_ptr = new TreeNode;
          t_ptr->left = subtree_l_ptr;
          t_ptr->right = subtree_r_ptr;
          res.push_back(t_ptr);
        }
      }
    }

    n_to_treenode_[n] = res;
    return res;
  }
};
